unify_idents/engine_parsers/ident/msfragger_3_parser.py

```diff
@@ -0,0 +1,207 @@
+
+        Args:
+            file (str): path to input file
+
+        Returns:
+            bool: True if parser and file are compatible
+
+        """
+        is_tsv = file.as_posix().endswith(".tsv")
+        with open(file.as_posix()) as f:
+            head = "".join([next(f) for _ in range(1)])
+        head = set(head.rstrip("\n").split("\t"))
+        ref_columns = {
+            "scannum",
+            "peptide",
+            "charge",
+            "peptide_prev_aa",
+            "peptide_next_aa",
+            "protein",
+            "modification_info",
+            "retention_time",
+            "precursor_neutral_mass",
+            "calc_neutral_pep_mass",
+            "hit_rank",
+            "massdiff",
+            "num_matched_ions",
+            "tot_num_ions",
+            "hyperscore",
+            "nextscore",
+            "num_tol_term",
+            "num_missed_cleavages",
+            "expectscore",
+            "best_locs",
+            "score_without_delta_mass",
+            "best_score_with_delta_mass",
+            "second_best_score_with_delta_mass",
+            "delta_score",
+        }
+        columns_match = len(ref_columns.difference(head)) == 0
+        return is_tsv and columns_match
+
+    def _map_mod_translation(self, row, map_dict):
+        """Replace single mod string.
+
+        Args:
+            row (str): unprocessed modification string
+            map_dict (dict): mod mapping dict
+
+        Returns:
+            mod_str (str): formatted modification string
+        """
+        mod_str = ""
+        if row == "" or row == [""]:
+            return mod_str
+        for mod in row:
+            mass = re.search(r"\(([^)]+)", mod).group(1)
+            name = map_dict[mass]
+            if len(name) > 0:
+                for m in name:
+                    if any(["N-term" in p for p in self.mod_dict[m]["position"]]):
+                        pos = 0
+                    else:
+                        pos = int(re.search(r"^\d+", mod).group(0))
+                    mod_str += f"{m}:{pos};"
+            else:
+                return "NON_MAPPABLE"
+        return mod_str
+
+    def translate_mods(self):
+        """
+        Replace internal modification nomenclature with formatted modification strings.
+
+        Returns:
+            (pd.Series): column with formatted mod strings
+        """
+        mod_split_col = self.df["Modifications"].fillna("").str.split(", ")
+        unique_mods = set().union(*mod_split_col.apply(set)).difference({""})
+        unique_mod_masses = {re.search(r"\(([^)]+)", m).group(1) for m in unique_mods}
+        # Map single mods
+        potential_names = {
+            m: [
+                name
+                for name in self.mod_mapper.mass_to_names(round(float(m), 4), decimals=4)
+                if name in self.mod_dict
+            ]
+            for m in unique_mod_masses
+        }
+        # Map multiple mods
+        for n in [2, 3]:
```

Comment:
this will create all potential 2 and 3 mass combos .. maybe start with 2 and if you could not map all mods make three ..?

Reply:
it does just that. see the next line. if potential_names value was already filled in during the n=2 iteration, the l.132 dict comprehension will be empty for n=3 and will be skipped.